core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -82,48 +100,157 @@ public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardR
+                indexMetaDataBuilder = updatePrimaryTerm(oldIndexMetaData, indexMetaDataBuilder, shardId, updates);
+            }
+
+            if (indexMetaDataBuilder != null) {
+                if (metaDataBuilder == null) {
+                    metaDataBuilder = MetaData.builder(oldMetaData);
+                }
+                metaDataBuilder.put(indexMetaDataBuilder);
+            }
+        }
+
+        if (metaDataBuilder != null) {
+            return metaDataBuilder.build();
+        } else {
+            return oldMetaData;
+        }
+    }
+
+    /**
+     * Updates in-sync allocations with routing changes that were made to the routing table.
+     */
+    private IndexMetaData.Builder updateInSyncAllocations(RoutingTable newRoutingTable, IndexMetaData oldIndexMetaData,
+                                                          IndexMetaData.Builder indexMetaDataBuilder, ShardId shardId, Updates updates) {
+        Set<String> oldInSyncAllocationIds = oldIndexMetaData.inSyncAllocationIds(shardId.id());
+
+        // check if we have been force-initializing an empty primary or a stale primary
+        if (updates.initializedPrimary != null && oldInSyncAllocationIds.isEmpty() == false &&
+            oldInSyncAllocationIds.contains(updates.initializedPrimary.allocationId().getId()) == false) {
+            // we're not reusing an existing in-sync allocation id to initialize a primary, which means that we're either force-allocating
+            // an empty or a stale primary (see AllocateEmptyPrimaryAllocationCommand or AllocateStalePrimaryAllocationCommand).
+            boolean emptyPrimary = updates.initializedPrimary.unassignedInfo().getReason() == UnassignedInfo.Reason.INDEX_CREATED;
+            assert updates.addedAllocationIds.isEmpty() : (emptyPrimary ? "empty" : "stale") +
+                " primary is not force-initialized in same allocation round where shards are started";
+
+            if (indexMetaDataBuilder == null) {
+                indexMetaDataBuilder = IndexMetaData.builder(oldIndexMetaData);
+            }
+            if (emptyPrimary) {
+                // forcing an empty primary resets the in-sync allocations to the empty set (ShardRouting.allocatedPostIndexCreate)
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(), Collections.emptySet());
+            } else {
+                // forcing a stale primary resets the in-sync allocations to the singleton set with the stale id
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(),
+                    Collections.singleton(updates.initializedPrimary.allocationId().getId()));
+            }
+        } else {
+            // standard path for updating in-sync ids
+            Set<String> inSyncAllocationIds = new HashSet<>(oldInSyncAllocationIds);
+            inSyncAllocationIds.addAll(updates.addedAllocationIds);
+            inSyncAllocationIds.removeAll(updates.removedAllocationIds);
```

Comment:
yes, that's already asserted before calling this method: https://github.com/elastic/elasticsearch/pull/20023/files/18bf5c435d0349c8da23a4f5e3523eec1646d041#diff-ff816fc88b23166c4be9374415818aceR123

I'll put it at the beginning of this method.